File: backend/app/services/phone.py
import re


def normalize_panama_phone(value: str) -> str | None:
    digits = re.sub(r"\D", "", value or "")
    if not digits:
        return None

    if digits.startswith("00"):
        digits = re.sub(r"^00+", "", digits)

    while digits.startswith("507") and len(digits) > 8:
        digits = digits[3:]

    if len(digits) == 9 and digits.startswith("0"):
        digits = digits[1:]

    if not re.fullmatch(r"\d{8}", digits):
        return None

    if not re.fullmatch(r"[2-8]\d{7}", digits):
        return None

    return f"+507{digits}"


def snap_phone_digits(e164: str) -> str:
    return re.sub(r"\D", "", e164)

File: frontend/src/lib/phone.ts
/** Strip to digits and normalize to E.164 (+507XXXXXXXX). */
export function normalizePanamaPhone(value: string): string | null {
  let digits = (value ?? "").replace(/\D/g, "");
  if (!digits) return null;

  // International prefix: 00507…
  if (digits.startsWith("00")) {
    digits = digits.replace(/^00+/, "");
  }

  // Country code repeated or prefixed once
  while (digits.startsWith("507") && digits.length > 8) {
    digits = digits.slice(3);
  }

  // Leading 0 before local number (e.g. 060012345)
  if (digits.length === 9 && digits.startsWith("0")) {
    digits = digits.slice(1);
  }

  if (!/^\d{8}$/.test(digits)) {
    return null;
  }

  // Panama national numbers start with 2–8 (mobile 6xxx-xxxx, landline 2–4, etc.)
  if (!/^[2-8]\d{7}$/.test(digits)) {
    return null;
  }

  return `+507${digits}`;
}

export function formatPanamaPhoneDisplay(value: string): string {
  const normalized = normalizePanamaPhone(value);
  if (!normalized) {
    const digits = value.replace(/\D/g, "").replace(/^507/, "").slice(0, 8);
    if (digits.length <= 4) return digits;
    return `${digits.slice(0, 4)}-${digits.slice(4)}`;
  }
  const local = normalized.slice(4);
  return `${local.slice(0, 4)}-${local.slice(4)}`;
}

export function createEventId(eventName: string) {
  return `${eventName}_${Date.now()}_${crypto.randomUUID()}`;
}
